### ISSUE TITLE
[FEATURE] Literal type support

### ISSUE THREAD

**Issue by alippai** [fastapi/typer#76]
Support for the literal type (python 3.8 or via typing_extensions) would be nice. Eg.
```python
def get_list(output: Literal["raw", "json", "yaml"] = "raw"):
  if output == "json":
    typer.echo(json.dumps([1, 2, 3]))
```

The only workarounds (using string or multiple bool flags + validating input) are ugly now.

**Comment by tiangolo** [comment]
I think you can solve it with an enum.

Here are the docs: https://typer.tiangolo.com/tutorial/parameter-types/enum/

**Comment by alippai** [comment]
This could work indeed

**Comment by tiangolo** [comment]
Great! If that answers your question you can then close the issue.

**Comment by Aran-Fey** [comment]
I don't get why this is closed. Does this mean you won't be adding support for `typing.Literal`? Or did the bot mess up?

**Comment by arogozhnikov** [comment]
Literal is pretty useful, and supported by pydantic for a while. 

Certainly vote for adding support. There is a pretty well-written PR too #174.

If that one isn't satisfactory, @tiangolo please state this, because reason for not supporting is unclear

**Comment by f4hy** [comment]
While things could be rewritten as an enum, it does not play nice with existing typed code which uses typing.Literal to define a set of allowed values. I hope we can revisit this.

**Comment by kimothy** [comment]
I agree as @f4hy as well. In many cases typing.Literal is a much more elegant solution then using enum.Enum.

**Comment by lainisourgod** [comment]
@tiangolo I have a case similar to @f4hy. In my ML project I use smth like `TargetType = typing.Literal["age", "income", "gender"]` to specify which targets I train model for, and use it in dicts (`{"age": 40, "income": 11}`), keys in exported json-s, and as argument types to functions so I can be sure I won't use unspecified target in my code.

I don't wanna rewrite everything else in enum style, it will be tedious and harder to use.

**Comment by danielgafni** [comment]
+1, this would be useful

**Comment by Yarden-zamir** [comment]
Can we have this reopened maybe? Enums are a different usecase. Yes they function the same in the end but they are ergonomically different. Using an enum is good for when the same slection of choices is used somewhere else too, and using a literal is useful when it's a one time thing and it helps show in a very clear way what options are possible in the same line
```python
def make_cheese(type: CheeseType)
```
This makes sense for an enum because CheeseType is something that will be useful in more places in the code.
```python
def make_food(for_person: AllPersonsICareAbout)
```
This makes less sense, because in my case, I don't care about who I can cook for in other places (for now)
```python
def make_food(for_person: Literal['Yarden', 'Bob']):
```
Either way, even if you disagree, the language evidently does agree, meaning those two are different and are both implemented so the argument that they are the same is not passing. You can argue that they are similar enough which makes it none critical to invest resources in implementing this now. I can agree with that, but if that's the case, reopen the issue and have it in the backlog, maybe someone else will implement it

**Comment by samsja** [comment]
I could use this as well :smiling_face_with_tear:

**Comment by dorinclisu** [comment]
Please reopen.

**Comment by Yarden-zamir** [comment]
This issue was closed 4 years ago. Maybe it's time for a new issue or fork?

**Comment by 5j9** [comment]
> This issue was closed 4 years ago. Maybe it's time for a new issue or fork?

[Cyclopts ](https://github.com/BrianPugh/cyclopts) supports this feature, see: https://cyclopts.readthedocs.io/en/latest/vs_typer/choices/README.html

**Comment by Kludex** [comment]
This is easy to implement. I'm reopening this issue.

**Comment by samsja** [comment]
yes please, I moved away from typer to cyclopts because Literal is so much easier than enum for cli

**Comment by pmeier** [comment]
Relevant PR #429 was approved. @Kludex can you share any timeline on when this will be merged and released?

**Comment by fzyzcjy** [comment]
Hi, is there any updates? Thanks!

**Comment by sector119** [comment]
But list of literals still doesn't work ?

**Comment by KelvinChung2000** [comment]
Is it not working now? Is it now from `typing` instead of `typing_extension`?
